### PR TITLE
[DSHARE-977] L-05 Update balanceToShares with FixedPointMathLib

### PR DESCRIPTION
**Pull request by joshualyguessennd**
Addresses rounding from Pashov review

> Function burn() in UsdPlus contract, burns specified USD+ tokens from caller. It has been used during the withdraw process.
> 
>     /// @notice burn USD+ from msg.sender
>     function burn(uint256 value) external {
>         address from = _msgSender();
>         _useBurningLimits(from, value);
>         _burn(from, value);
>     }
> Because USD+ is a rebasing token to calculate the share burn amount code use _balancePerShare and divide the USD+ amount by this value.
> 
>     function balanceToShares(uint256 balance) public view returns (uint256) {
>         return mulDiv(balance, _INITIAL_BALANCE_PER_SHARE, balancePerShare()); // floor
>     }
> The issue is that by doing this the number of share amount would be round down and as result code would burn less than what caller specified. This would cause small profit for caller and in some circumstances malicious user can extract value because of this.
> 
> Round up in favor of contract when calculating the share amount in burn.

I (@jaketimothy ) added rounding change to transfer and burn.

### CODE REVIEW

File: src/ERC20Rebasing.sol [hunk 1]
```diff
@@ -27,7 +28,7 @@ abstract contract ERC20Rebasing is ERC20 {
     }
 
     function balanceToShares(uint256 balance) public view returns (uint256) {
-        return mulDiv(balance, _INITIAL_BALANCE_PER_SHARE, balancePerShare()); // floor
+        return FixedPointMathLib.mulDiv(balance, _INITIAL_BALANCE_PER_SHARE, balancePerShare()); // floor
```
Comment (jaketimothy) [inline src/ERC20Rebasing.sol:31]: Change comment to "ceil". Use correct method - this one rounds down.

**Review by jaketimothy — CHANGES_REQUESTED** [review]
You will probably need to use fullMulDivUp to pass tests and round correctly

**Comment by jaketimothy** [comment]
Can you also replace all usages of prb-math in this file and remove that import? I think we should be consistent

File: src/ERC20Rebasing.sol [hunk 2]
```diff
@@ -39,9 +40,9 @@ abstract contract ERC20Rebasing is ERC20 {
     function maxSupply() public view virtual returns (uint256) {
         uint128 balancePerShare_ = balancePerShare();
         if (balancePerShare_ < _INITIAL_BALANCE_PER_SHARE) {
-            return mulDiv18(type(uint256).max, balancePerShare_);
+            return FixedPointMathLib.mulWad(type(uint256).max, balancePerShare_);
```
Comment (jaketimothy) [inline src/ERC20Rebasing.sol:43]: Can you add a test for this?

File: src/ERC20Rebasing.sol [hunk 3]
```diff
@@ -23,11 +23,12 @@ abstract contract ERC20Rebasing is ERC20 {
     function balancePerShare() public view virtual returns (uint128);
 
     function sharesToBalance(uint256 shares) public view returns (uint256) {
-        return mulDiv18(shares, balancePerShare()); // floor
+        // WAD = 1e18
+        return FixedPointMathLib.fullMulDiv(shares, balancePerShare(), 1e18); // floor
```
Comment (jaketimothy) [inline src/ERC20Rebasing.sol:27]: Does mulWad work here or overflow?

**Review by jaketimothy — CHANGES_REQUESTED** [review]

File: src/ERC20Rebasing.sol (hunk 3 above)
Comment (joshualyguessennd) [reply src/ERC20Rebasing.sol:27]: Overflow

Comment (jaketimothy) [reply src/ERC20Rebasing.sol:27]: Then the other use of mulWad should be changed to fullMulDiv

File: src/ERC20Rebasing.sol [hunk 4]
```diff
@@ -23,11 +23,12 @@ abstract contract ERC20Rebasing is ERC20 {
...
     }
 
     function balanceToShares(uint256 balance) public view returns (uint256) {
-        return mulDiv(balance, _INITIAL_BALANCE_PER_SHARE, balancePerShare()); // floor
+        return FixedPointMathLib.fullMulDiv(balance, _INITIAL_BALANCE_PER_SHARE, balancePerShare()); // ceil
```
Comment (jaketimothy) [inline src/ERC20Rebasing.sol:31]: This should be fullMulDivUp

**Review by jaketimothy — CHANGES_REQUESTED** [review]

File: src/ERC20Rebasing.sol (hunk 4 above)
Comment (joshualyguessennd) [reply src/ERC20Rebasing.sol:31]: It failed with fullMulDivUp

Comment (jaketimothy) [reply src/ERC20Rebasing.sol:31]: Investigating

**Review by jaketimothy — DISMISSED** [review]

File: src/ERC20Rebasing.sol [hunk 5]
```diff
@@ -101,20 +106,23 @@ abstract contract ERC20Rebasing is ERC20 {
...
         }
+        // Check overflow
+        if (totalSharesAfter < totalSharesBefore) revert TotalSupplyOverflow();
+        // Check total supply limit
+        // Same as maxSupply logic, other balancePerShare scenario covered by the previous check
+        uint128 balancePerShare_ = balancePerShare();
+        if (
+            balancePerShare_ > _INITIAL_BALANCE_PER_SHARE
+                && sharesToBalance(totalSharesAfter)
+                    > FixedPointMathLib.fullMulDiv(type(uint256).max, _INITIAL_BALANCE_PER_SHARE, balancePerShare_)
```
Comment (etet245) [inline src/ERC20Rebasing.sol:124]: is there a reason for spelling this out over calling `maxSupply()`?

**Review by etet245 — APPROVED** [review]

**Review by jaketimothy — APPROVED** [review]

**Pull request merged**